onnxruntime/core/common/threadpool.cc

```diff
@@ -26,6 +26,82 @@ namespace onnxruntime {
 
 namespace concurrency {
 
+ThreadPoolProfiler::ThreadPoolProfiler() {
+  memset(events_, 0, sizeof(uint64_t) * MAX_EVENT);
+}
+
+bool ThreadPoolProfiler::Enabled() const {
+  return enabled_ && std::this_thread::get_id() == main_thread_id_;
```

Comment:
>enabled_ [](start = 9, length = 8)

is enabled_ needed? is the same information available with `main_thread_id_ != std::thread::id{}`?
should main_thread_id_ be atomic? it looks like multiple threads may access it.
would starting_thread_id_ be a better name for main_thread_id_? i.e., the id of the thread that called Start().

Reply:
Is it possible to make it inlined?

Reply:
For now the profiler is supposed to be only consumed in sequential executor, namely the main thread that launches each op, meaning the Start() and Stop() functions are not designed for concurrent access . 
enable_ is here to quickly exclude threads from logging the numbers by function LogStart(), LogEnd() and LogEndAndStart() when the session is not in profiling mode, thereby reduce the overhead.

Reply:
Inline is not a bad one btw.

Reply:
is it true that enabled_ can be replaced with main_thread_id_ != thread::id{}? that check could still be done before main_thread_id_ != this_thread::get_id().

---
In reply to: [585087214](https://github.com/microsoft/onnxruntime/pull/6748#discussion_r585087214) [](ancestors = 585087214)